pkg/chains/formats/slsa/v2alpha2/internal/resolved_dependencies/resolved_dependencies.go

```diff
@@ -0,0 +1,361 @@
+			return err
+		}
+	}
+	return nil
+}
+
+// addSidecarImagesToResolvedDependencies adds sidecar images to predicate.resolvedDependencies
+func addSidecarImagesToResolvedDependencies(sidecars []v1beta1.SidecarState, resolvedDependencies *[]v1.ResourceDescriptor) error {
+	for _, sidecarState := range sidecars {
+		if err := addImageIDToResolvedDependencies(sidecarState.ImageID, resolvedDependencies); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// addImageIDToResolvedDependencies converts an imageId with format <uri>@sha256:<digest> and then adds it to a provenance resolvedDependencies.
+func addImageIDToResolvedDependencies(imageID string, resolvedDependencies *[]v1.ResourceDescriptor) error {
+	rd := v1.ResourceDescriptor{
+		Digest: common.DigestSet{},
+	}
+	uriDigest, err := material.ExtractUriDigestFromImageID(imageID)
+	if err != nil {
+		return err
+	}
+	rd.URI = uriDigest["uri"]
+	rd.Digest[uriDigest["digestAlgorithm"]] = uriDigest["digestValue"]
+	rd.Name = containerImageName
+	*resolvedDependencies = append(*resolvedDependencies, rd)
+	return nil
+}
+
+// gitInfo scans over the input parameters and looks for parameters
+// with specified names.
+func gitInfo(tro *objects.TaskRunObject) (commit string, url string) {
+	// Scan for git params to use for resolved dependencies
+	if tro.Status.TaskSpec != nil {
+		for _, p := range tro.Status.TaskSpec.Params {
+			if p.Default == nil {
+				continue
+			}
+			if p.Name == attest.CommitParam {
+				commit = p.Default.StringVal
+				continue
+			}
+			if p.Name == attest.URLParam {
+				url = p.Default.StringVal
+			}
+		}
+	}
+
+	for _, p := range tro.Spec.Params {
+		if p.Name == attest.CommitParam {
+			commit = p.Value.StringVal
+			continue
+		}
+		if p.Name == attest.URLParam {
+			url = p.Value.StringVal
+		}
+	}
+
+	for _, r := range tro.Status.TaskRunResults {
+		if r.Name == attest.CommitParam {
+			commit = r.Value.StringVal
+		}
+		if r.Name == attest.URLParam {
+			url = r.Value.StringVal
+		}
+	}
+
+	url = attest.SPDXGit(url, "")
+	return
+}
+
+// removeDuplicateResolvedDependencies removes duplicate resolved dependencies from the slice of resolved dependencies.
+// Original order of resolved dependencies is retained.
+func removeDuplicateResolvedDependencies(resolvedDependencies []v1.ResourceDescriptor) ([]v1.ResourceDescriptor, error) {
+	out := make([]v1.ResourceDescriptor, 0, len(resolvedDependencies))
+
+	// make map to store seen resolved dependencies
+	seen := map[string]bool{}
+	for _, resolvedDependency := range resolvedDependencies {
+		rDep := v1.ResourceDescriptor{}
+		rDep.URI = resolvedDependency.URI
+		rDep.Digest = resolvedDependency.Digest
+		rd, err := json.Marshal(rDep)
+		if err != nil {
+			return nil, err
+		}
+		if seen[string(rd)] {
+			// if the resource is a top level task or pipeline config then add it regardless.
+			if resolvedDependency.Name == taskConfigName || resolvedDependency.Name == pipelineConfigName {
```

Comment:
Not quite following why does this need to be special cased. Can you add more detail?

Reply:
Here, I wanted to ensure that we added the top level pipeline and task config to resolved dependencies. In `slsav0.2`, we had invocation.configSource for that but we don't have that here.

I wanted to ensure that the top level `task/pipeline` config is included (instead of being ignored) even if it is a repeated dependency.